protos/mocap/mocap.proto

```diff
@@ -0,0 +1,49 @@
+syntax = "proto3";
+
+package mavsdk.rpc.mocap;
+
+option java_package = "io.mavsdk.mocap";
+option java_outer_classname = "MocapProto";
+
+/**
+ * TODO
+ */
+service MocapService {
+    /*
+     * TODO
+     */
+    rpc CreateVisionPositionEstimateStream(stream VisionPositionEstimateRequest) returns(stream VisionPositionEstimateResponse) {}
+}
+
+message VisionPositionEstimateRequest {
+    VisionPositionEstimate vision_position_estimate = 1;
+}
+message VisionPositionEstimateResponse {
+  MocapResult mocap_result = 1;
+}
+
+// Global position/attitude estimate from a vision source.
+message VisionPositionEstimate {
+    float x = 1; // Global X position (m)
+    float y = 2; // Global Y position (m)
+    float z = 3; // Global Z position (m)
+    float roll = 4; // Roll angle (rad)
+    float pitch = 5; // Pitch angle (rad)
+    float yaw = 6; // Yaw angle (rad)
+    repeated float covariance = 7; // Row-major representation of pose 6x6 cross-covariance matrix upper right triangle
+    uint32 reset_counter = 8; // Estimate reset counter.
+}
+
+// Result type.
+message MocapResult {
```

Comment:
I don't know the Mocap interface. Does that have to be a stream? Sounds like it gives the state regarding "creating a position estimate stream from the client", doesn't it?

Reply:
> creating a position estimate stream from the client
Yes, it is
I think that stream is more suitable for these purposes. Position estimate frame should have at least 50Hz frequency, so I think that Stream exactly what is needed.

Reply:
But on the other side if we want to know position set request result may be will be more suitable single request-response scheme. 
But may be that will be slowly than stream?

Reply:
I'm not sure I get that. Is it normal to have a stream of `MocapResult`? When do I receive `SUCCESS`? When I open the connection, to tell me that I can now start sending values? Every time I send a new value? When I'm done sending values and the connection was closed successfully?

Reply:
Ok I think I get it: we continuously send new position estimates, and that can fail if the connection somehow fails. But it's not like PX4 sends a `SUCCESS` status for each message. We just assume that if we could send the message, then it is a `SUCCESS`.

Until now, receiving `SUCCESS` from the server has always meant "end of stream". The only one having "intermediate results" is the [calibration plugin](https://github.com/mavlink/MAVSDK-Proto/blob/master/protos/calibration/calibration.proto#L54-L56), which has `IN_PROGRESS` and `INSTRUCTION` (that I believe should be merged into a common `NEXT` message). So that we would have:

- `SUCCESS` means that the process is complete
- `NEXT` is an intermediate message
- Everything else is a failure

And we could consider using `NEXT` here as well. But I'm tempted to think that in this case, we actually only want to act on error, i.e. the server-side stream would not send anything unless it is an error. We could have a result for that, e.g. `OK` or `NOOP`, that would be returned by the C++ implementation, but that would be ignored by gRPC...

Not sure if what I'm saying is clear :sweat_smile:. Does it make any sense at all?

Reply:
> We just assume that if we could send the message, then it is a `SUCCESS`.

So we need to define "normal" execution without response.
In that case we should have something like 
`rpc VpeStream (stream VpeRequest) returns(VpeResponse)`
But in gRPC we cant cancel client stream on server-side (with streaming from client to server) **and send error code as response**. May be I wrong, but I [don't find](https://grpc.io/docs/guides/error/) way to close client stream on server-side and send response code (like `VpeResponse`).

I suggest this solution:

 - RPC definition:
`rpc VpeStream(stream VpeRequest)  returns(stream VpeResponse)`

 - If received `VpeRequest` sent to UAV without errors - do not respond.

 - If received `VpeRequest` cant be send to UAV (connection lost or any other problem) - send to client `VpeResponse` with error code

 - After first `VpeResponse` code was sent - reply corresponding error on every next `VpeRequest` until connection restore

 - If connection has been restored - send `VpeResponse::SUCCESS" once and after that continue without `VpeResponse` until next communication error.

Reply:
>  I think that the connection will be closed as soon as the server sends the response (needs to be confirmed).

Hm. I will check that. If it really is so it is what we need, yes.

> Anyway if you lose the connection to the system, all the messages you send from MAVDSK will fail.

Need to confirm. 
I perform one small test on MAVSDK. Connect (over PixRacer WiFi), disconnect (turn WiFi off) and connect again (turn on WiFi). Seems that backend can regain connection:
```
[03:07:48|Debug] Sending msg 0 (system_impl.cpp:495)
[03:07:49|Debug] Sending msg 0 (system_impl.cpp:495)
[03:07:50|Debug] Sending msg 0 (system_impl.cpp:495)
[03:07:51|Info ] heartbeats timed out (system_impl.cpp:305) # Disconnect
[03:07:51|Debug] Lost 3546392297223829815 (mavsdk_impl.cpp:388)
[03:07:51|Info ] System timed out [UUID: 3546392297223829815] (connection_initiator.h:39)
[03:08:03|Debug] Discovered 2 component(s) (UUID: 3546392297223829815) (system_impl.cpp:558)
[03:08:03|Debug] Sending msg 76 (system_impl.cpp:495)
[03:08:03|Debug] Sending msg 76 (system_impl.cpp:495)
[03:08:03|Debug] Sending msg 0 (system_impl.cpp:495)
[03:08:03|Debug] MAVLink: critical: Data link lost (system_impl.cpp:300)
[03:08:03|Debug] MAVLink: info: Data link regained (system_impl.cpp:300)  # Connect
[03:08:04|Debug] Sending msg 0 (system_impl.cpp:495)
```

> Is it important to know which one failed? I understand it more as some kind of unreliable channel: you always want the server to know the latest message, but you don't care about the ones that were missed. Isn't that correct?

I think that you are right. There is no reason to know which exactly message was lost.

Reply:
> Hm. I will check that. If it really is so it is what we need, yes.

Did you find time to confirm that? Would be helpful :-)

Reply:
Seems like we cant close stream with Server side and send Response.

We only can use 

`response->set_allocated_mocap_result(response_val);`

This result will be sent when client call [`writer->WritesDone();` ](https://grpc.io/docs/tutorials/basic/cpp/).
So at this moment I think that most easy and clear way to implement that is both-side-streaming.

Reply:
> Seems like we cant close stream on Server side with sending Response.

I don't understand what that means. When the server sends `return grpc::Status::OK;` in your C++ code, it most likely closes the client-side stream. Right?

Reply:
At this time I think that both-side streaming too complicated because of python-generator cant generate appropriate code. 
Returned to single Request-Response scheme (https://github.com/mavlink/MAVSDK-Proto/pull/105#issuecomment-534513268).
It is a pity that such a long discussion could not be used so far :(
I will check that if will necessary to come back to client-server streaming in this module